src/traces/waterfall/calc.js

```diff
@@ -0,0 +1,123 @@
+/**
+* Copyright 2012-2019, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var Axes = require('../../plots/cartesian/axes');
+var hasColorscale = require('../../components/colorscale/helpers').hasColorscale;
+var colorscaleCalc = require('../../components/colorscale/calc');
+var arraysToCalcdata = require('./arrays_to_calcdata');
+var calcSelection = require('../scatter/calc_selection');
+
+function extractOperators(list) {
```

Comment:
Interesting, but I can't allow. Again, this would make `waterfall` data array "too different" from our other trace types.

What's wrong with having a separate data array e.g. `sums` as proposed in https://github.com/plotly/plotly.js/issues/3037#issuecomment-454864262 ?

Reply:
Using a pattern like this https://github.com/plotly/plotly.js/blob/b0fdceda91abdbf2ccfccde1b4106db208931ea5/test/image/mocks/waterfall_profit-loss_2018_colorscale.json#L7-L23 we no longer use `magic` numbers and it may be easier for the users to add and figure out the calculation related for each row.
But yes I see your point @etpinard and would likely switch to use an array for the operations there.

Reply:
Revised in https://github.com/plotly/plotly.js/pull/3531/commits/4383a71e7f52c8ac4aeb3313cc44d3f3146a929c.